README
/*
    IDA Pro Collabreation/Synchronization Plugin
    Copyright (C) 2018 Chris Eagle <cseagle at gmail d0t com>
    Copyright (C) 2018 Tim Vidas <tvidas at gmail d0t com>


    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the Free
    Software Foundation; either version 2 of the License, or (at your option)
    any later version.

    This program is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
    FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
    more details.

    You should have received a copy of the GNU General Public License along with
    this program; if not, write to the Free Software Foundation, Inc., 59 Temple
    Place, Suite 330, Boston, MA 02111-1307 USA

*/

Version 1.0.0

BACKGROUND:

IdaPro has an event notification mechanism that plugins may utilize to be
notified of many different types of changes being made to a database. The
primary idea behind CollabREate is to hook these change notifications and
broadcast changes to a central database server that records each change 
and forwards the associated change parameters to other interested users
working on the same binary file. The theory being that one user's changes
will be applied to a second user's database effectively allowing multiple
users to simultaneously edit and annotate their databases.

Please not that no single master copy of a particular IDB is maintained
anywhere. A central database that records all events generated by all
registered users is the only record of changes. It is entirely possible
that the state of two databases diverges if one user were to make
modifications without publishing those modifications to the database server.

A major difficulty in this approach arises from the manner in which Ida
notifications are generated. Because of the manner in which the Ida auto
analysis subsystem operates, what appears to be a single action to the user
may in fact translated into many (tens, hundreds, or even thousands) 
individual notification messages which all get pushed to the database. This
makes conflict resolution between users and events extremely difficult
because it is virtually impossible to associate a specific notification
message with a specific user action. In fact some notification messages may,
at first glance seem entirely unrelated to the user action which resulted in 
the generation of those notifications.

Another challenge that arises is to supress the generation of duplicate 
notifications when processing incoming notifications. The current version of
CollabREate simply de-registers itself for notifications while applying any
changes indicated by each incoming message. This seems to work well unless
the incoming messages results in the generation of more than one additional
notification in the receiving IDB. In such cases, it is extremely difficult
to determine when the last notification associated with the application of
a newly received message has be generated, and therefore difficult to 
determine exactly when to re-register for receipt of local notifications.
This is suspected to be one of the primary causes of a CollabREate database
getting flooded with notifications and associated IDBs diverging in state.
In practice, this problem seems to arise from actions that generate large
numbers of notifications such as undefining large blocks of code or data
or patching large numbers of bytes.

RELEASE NOTES version 1.0.0

Removed all Qt dependeicies, changed to use IDA API user interface functions.
Notification protocol completely changed from a binary protocol to a
json based protocol. Server and database updated accordingly.
Support for IDA versions 6.5 through 7.0, older version no longer supported.
Binaries now land in bin directory

RELEASE NOTES version 0.4.0

Removed dependency on QtNetwork. Reworked asynchronous networking
for IDA versions 5.5 and later to make use of the IDA SDK execute_sync
function. Updated MSVC build files so that a single build file accomodates
multiple versions of IDA.

RELEASE NOTES version 0.3.0

This version of the plugin is NOT compatible with older (< 0.3.0) collabreate
databases. This is a result of the addtion of 64 bit support.  If you need
to use old databases, then you should continue to use version 0.2.0. We may
publish an upgrade utility that will upgrade 0.2.0 databases to 0.3.0 databases.
For the purposes of this discussion we are talking about the SQL database that
stores all of the IDA update notifications.

BUILDING THE PLUGIN

The IDA Pro plugin requires the IDA Pro SDK, which is distributed along with
Pro.  You need the version of the SDK that matches the version of IDA with 
which you intend to use the plugin.

First you clone the repo, you should do this within
the SDK's "plugins" directory:

cd <SDKDIR>/plugins
git clone https://github.com/cseagle/collabREate.git

BUILDING FOR WINDOWS

Using Visual Studio 2013 or later, open collabrete.sln

Based on your IDA version select and build the corresponding project

IDA Version         Project to build       Output file
<= 6.95 idaq        Release / Win32        bin/collabreate.plw
<= 6.95 idaq64      Release64 / Win32      bin/collabreate.p64
>= 7.0 ida          Release / x64          bin/collabreate.dll
>= 7.0 ida64        Release64 / x64        bin/collabreate64.dll

BUILDING FOR LINUX / OSX

Use the supplied Makefile to build the plugin

$ make

You may need to adjust paths to your SDK an IDA install in the Makefile
see IDA_SDK and IDAHOME in the Makefile.

Compiled binairs should land in the bin directory.

The plugin depends on the json-c library from github:
https://github.com/json-c/json-c
NOTE that the plugin uses features that are not available in the older
versions of json-c that install with most package managers, so build
and install from the github sources. 


INSTALLING THE PLUGIN

As with any IDA plugin, simply copy the compiled plugin file into collabreate's
bin directory to <idadir>/plugins, where <idadir> represents the location in
which IDA is installed.

Due to Windows' file locking, you will likely need to make sure all Ida databases
are closed before copying to this directory.

USING THE PLUGIN

It is intended that a new collabreate session is used for each binary loaded
into IDA Pro.  Once a binary is loaded and the autoanalysis phase has completed
you can activate CollabREate by via the hotkey ALT-F6 or the Edit/Plugins menu.

Upon activation you will be presented with a series of dialogs prompting you to 
connect to a CollabREate server, authenticate, start a new project, etc.  The 
exact series of steps will very upon the running mode of the server (such as if
it is connected to a backend database or not).

Typical Project Join Actions:
 Connect:           Connects to a collabreate server (default port is 5042)
 Authenticate:      Provide a User / Password to connect to the server
 Choose Project:    Provides a list of 'related' projects (1)
  New Project:      Request that the server creates a new project (2)
  Existing Project: Join an existing project
  Project Snapshot: Request a new project that is a Fork of an existing project
                    at a particular, previously saved point (2)

(1) only projects that match the MD5 of the binary loaded in IDA are displayed
(2) when starting a new project you must provide a textual description

When connecting to an existing project for the first time, you automatically
recieve all updates that have been made to the project prior to you joining.
On subsequent connections you will automatically receive any updates that have
been made since the last time you were connected to the project.

Once connected to a project, you can use IDA normally.  Updates you make are
sent to the server, updates other collabreators make are reflected in your
database.


Reactivating the plugin while the plugin is already activated will bring up a menu of 
collabreate specific commands.

Collabreation Actions:
 Disconnect:        Disconnect from the server.
 Fork:              Immediatly fork the project with the updates that IDA has
                    recieved and join this new project. (3)
 Snapshot:          Save a snapshot (point in time) (3)
 Manage Req Perms:  Allows the user to change the permissions requested when 
                    joining the project
 Manage Proj Perms: Allows the user (only if the user is the owner of the 
                    project) to alter the project permissions.

(3) the user must provide a textual description

Forking creates a new projet that can have further updates, and that others can
participate in (join).  Snapshots are simply 'projects at a point in time' - not
actual projects.  You can not 'join' a snapshot.  You must fork a new project
from a snapshot.  Overall, snapshots take up much less space than forked 
projects.


If the server is connected to a database backend, a unique project id is stored
in the idb file and subsequent connections to a server will automatically 
connect to the correct project.


ChangeLog
* 01/13/18
  Migrate to json based protocol. Old protocol and database form no longer supported.
  Supporting IDA versions 6.5-7.0
  Binaries now land in bin directory

* 01/02/17
  Conversion to protocol version 4, a JSON based messaging protocol
  Only the C++ server has been updated at this point
  Attempt to use only native IDA api functions for user interface
  Eliminate QT specific code to make building easier

* 05/25/14
  Added primitive messaging capability in CollabREate window. Message history is
  cached in the IDB and restored with each session.
  Added primitive support for working offline. If IDB has been previously joined
  to a project, then all changes in subsequent sessions will be cached locally
  until reconnected to the project server.

* 10/04/13
  Added support for repeatable comments in struct/enum per suggestion of Nico Bendlin
  Added support for additional offset reference types in COMMAND_OP_TYPE_CHANGED per
  contribution of Willem Jan Palenstijn
  Fixed handling of enum member comments and struct member comments

* 10/01/13
  Incorporated patch contributions from Willem Jan Palenstijn
  
* 08/24/12
  Removed dependency on QtNetwork. Reworked asynchronous networking
  for IDA versions 5.5 and later to make use of the IDA SDK execute_sync
  function.

* 09/08/11
  Added support for ida64, released version 0.3.0

* 10/20/10
  Added message handler for MSG_PROJECT_SNAPSHOT_REPLY

* 10/12/10
  Completed port to Qt for Linux and OS X

* 10/08/10
  Added initial Qt interface for Ida >= 6.0
  Added ubuntu-init.d-script server startup script contributed by William Bartell
  Added NO_OBSOLETE_FUNCTIONS macro into the project
  
* 02/28/09
  Fixed buffer overflow problem pointed out by Jeremy Cooper

* 08/27/08
  Added local tracking of enum names to allow renaming of enums to work.
  Added more tests to narrow down the exact type being referenced when
  an operand type is changed.  Specifically enum and structure offsets.
  These changes will not fix enum and struct updates in existing project
  databases, however new enum and struct updates will be properly stored
  into those databases once you have upgraded you plugin.

* 08/24/08
  Fix to allow proper publishing of name changes that take place in the 
  .bss section

* 08/16/08
  added permission handling for COMMAND_AREA_CMT_CHANGED

* 08/15/08
  Initial release of version 1.0


collabreate.h
/*
    IDA Pro Collabreation/Synchronization Plugin
    Copyright (C) 2018 Chris Eagle <cseagle at gmail d0t com>
    Copyright (C) 2018 Tim Vidas <tvidas at gmail d0t com>


    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the Free
    Software Foundation; either version 2 of the License, or (at your option)
    any later version.

    This program is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
    FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
    more details.

    You should have received a copy of the GNU General Public License along with
    this program; if not, write to the Free Software Foundation, Inc., 59 Temple
    Place, Suite 330, Boston, MA 02111-1307 USA

*/
#ifndef __COLLABREATE_H__
#define __COLLABREATE_H__

#include <stdint.h>
#include <stdarg.h>
#include <time.h>
#include <json-c/json.h>

#define NO_OBSOLETE_FUNCS
#define USE_DANGEROUS_FUNCTIONS
#define USE_STANDARD_FILE_FUNCTIONS

#include <pro.h>

#define PLUGIN_NAME "collabREate"

#define PROTOCOL_VERSION             4

#define JSON_NEW_CONST_KEY (JSON_C_OBJECT_ADD_KEY_IS_NEW | JSON_C_OBJECT_KEY_IS_CONSTANT)

#define COMMAND_BYTE_PATCHED         "byte_patched"
#define COMMAND_CMT_CHANGED          "cmt_changed"
#define COMMAND_TI_CHANGED           "ti_changed"
#define COMMAND_OP_TI_CHANGED        "op_ti_changed"
#define COMMAND_OP_TYPE_CHANGED      "op_type_changed"
#define COMMAND_ENUM_CREATED         "enum_created"
#define COMMAND_ENUM_DELETED         "enum_deleted"
#define COMMAND_ENUM_BF_CHANGED      "enum_bf_changed"
#define COMMAND_ENUM_RENAMED         "enum_renamed"
#define COMMAND_ENUM_CMT_CHANGED     "enum_cmt_changed"
#define COMMAND_ENUM_CONST_CREATED   "enum_const_created"
#define COMMAND_ENUM_CONST_DELETED   "enum_const_deleted"
#define COMMAND_STRUC_CREATED        "struc_created"
#define COMMAND_STRUC_DELETED        "struc_deleted"
#define COMMAND_STRUC_RENAMED        "struc_renamed"
#define COMMAND_STRUC_EXPANDED       "struc_expanded"
#define COMMAND_STRUC_CMT_CHANGED    "struc_cmt_changed"

#define COMMAND_CREATE_STRUC_MEMBER_DATA   "create_struc_mbr_data"
#define COMMAND_CREATE_STRUC_MEMBER_STRUCT "create_struc_mbr_struc"
#define COMMAND_CREATE_STRUC_MEMBER_REF    "create_struc_mbr_ref"
#define COMMAND_CREATE_STRUC_MEMBER_STROFF "create_struc_mbr_stroff"
#define COMMAND_CREATE_STRUC_MEMBER_STR    "create_struc_mbr_str"
#define COMMAND_CREATE_STRUC_MEMBER_ENUM   "create_struc_mbr_enum"

#define COMMAND_STRUC_MEMBER_DELETED      "struc_mbr_deleted"

//#define COMMAND_STRUC_MEMBER_RENAMED
#define COMMAND_SET_STACK_VAR_NAME        "set_stack_var_name"
#define COMMAND_SET_STRUCT_MEMBER_NAME    "set_struc_mbr_name"

//#define COMMAND_STRUC_MEMBER_CHANGED
#define COMMAND_STRUC_MEMBER_CHANGED_DATA   "struc_mbr_chg_data"
#define COMMAND_STRUC_MEMBER_CHANGED_STRUCT "struc_mbr_chg_struc"
#define COMMAND_STRUC_MEMBER_CHANGED_STR    "struc_mbr_chg_str"

#define COMMAND_THUNK_CREATED        "thunk_created"
#define COMMAND_FUNC_TAIL_APPENDED   "func_tail_appended"
#define COMMAND_FUNC_TAIL_REMOVED    "func_tail_removed"
#define COMMAND_TAIL_OWNER_CHANGED   "tail_owner_chg"
#define COMMAND_FUNC_NORET_CHANGED   "func_noret_chg"
#define COMMAND_SEGM_ADDED           "segm_added"
#define COMMAND_SEGM_DELETED         "segm_deleted"
#define COMMAND_SEGM_START_CHANGED   "segm_start_chg"
#define COMMAND_SEGM_END_CHANGED     "segm_end_chg"
#define COMMAND_SEGM_MOVED           "segm_moved"
//#define COMMAND_AREA_CMT_CHANGED     "area_cmt_chg"
#define COMMAND_RANGE_CMT_CHANGED     "range_cmt_chg"
#define COMMAND_STRUC_MEMBER_CHANGED_OFFSET "struc_mbr_chg_offset"
#define COMMAND_STRUC_MEMBER_CHANGED_ENUM   "struc_mbr_chg_enum"
#define COMMAND_CREATE_STRUC_MEMBER_OFFSET  "create_struc_mbr_offset"

#define AREACB_FUNCS                  "funcs"
#define AREACB_SEGS                   "segs"
#define RANGE_FUNCS     AREACB_FUNCS
#define RANGE_SEGS      AREACB_SEGS

#define COMMAND_IDP                 128
#define COMMAND_UNDEFINE            "undefine"
#define COMMAND_MAKE_CODE           "make_code"
#define COMMAND_MAKE_DATA           "make_data"
#define COMMAND_MOVE_SEGM           "move_segm"
#define COMMAND_RENAMED             "renamed"
#define COMMAND_ADD_FUNC            "add_func"
#define COMMAND_DEL_FUNC            "del_func"
#define COMMAND_SET_FUNC_START      "set_func_start"
#define COMMAND_SET_FUNC_END        "set_func_end"
#define COMMAND_VALIDATE_FLIRT_FUNC "validate_flirt_func"
#define COMMAND_ADD_CREF            "add_cref"
#define COMMAND_ADD_DREF            "add_dref"
#define COMMAND_DEL_CREF            "del_cref"
#define COMMAND_DEL_DREF            "del_dref"

#define SERVER_MAP_TID              200
#define SERVER_RENAME_STRUCT        201
#define COMMAND_USER_MESSAGE        "user_message"

//all idb manipulation messages must be <= this number
#define MSG_IDA_MAX                 255

#define MSG_CONTROL_FIRST            "control_first"
#define MSG_INITIAL_CHALLENGE        "initial_challenge"
#define MSG_AUTH_REQUEST             "auth_request"
#define MSG_AUTH_REPLY               "auth_reply"
#define AUTH_REPLY_SUCCESS           1
#define AUTH_REPLY_FAIL              0
#define MSG_PROJECT_LIST             "project_list"
#define MSG_PROJECT_JOIN_REQUEST     "project_join_request"
#define MSG_PROJECT_JOIN_REPLY       "project_join_reply"
#define JOIN_REPLY_SUCCESS           1
#define JOIN_REPLY_FAIL              0
#define MSG_PROJECT_NEW_REQUEST      "project_new_request"
#define MSG_SEND_UPDATES             "send_updates"
#define MSG_PROJECT_REJOIN_REQUEST   "project_rejoin_request"
#define MSG_ACK_UPDATEID             "ack_updateid"
#define MSG_PROJECT_SNAPSHOT_REQUEST "project_snapshot_request"
#define MSG_PROJECT_SNAPSHOT_REPLY   "project_snapshot_reply"
#define PROJECT_SNAPSHOT_SUCCESS 1
#define PROJECT_SNAPSHOT_FAIL    0
#define MSG_PROJECT_FORK_REQUEST     "project_fork_request"
#define MSG_PROJECT_SNAPFORK_REQUEST "project_snapfork_request"
#define MSG_PROJECT_FORK_FOLLOW      "project_fork_follow"
#define MSG_PROJECT_LEAVE            "project_leave"
#define MSG_GET_REQ_PERMS            "get_req_perms"
#define MSG_GET_REQ_PERMS_REPLY      "get_req_perms_reply"
#define MSG_SET_REQ_PERMS            "set_req_perms"
#define MSG_SET_REQ_PERMS_REPLY      "set_req_perms_reply"
#define MSG_GET_PROJ_PERMS           "get_proj_perms"
#define MSG_GET_PROJ_PERMS_REPLY     "get_proj_perms_reply"
#define MSG_SET_PROJ_PERMS           "set_proj_perms"
#define MSG_SET_PROJ_PERMS_REPLY     "set_proj_perms_reply"

#define MSG_ERROR                    "collab_error"
#define MSG_FATAL                    "collab_fatal"

class netnode;
extern netnode cnn;
extern qstrvec_t msgHistory;
extern qstring *changeCache;

#define COLLABREATE_NETNODE "$ COLLABREATE NETNODE"

#define COLLABREATE_ENUMS_TAG 'E'
#define COLLABREATE_STRUCTS_TAG 'T'
#define COLLABREATE_MSGHISTORY_TAG ((char)0x81)
#define COLLABREATE_CACHE_TAG ((char)0x82)

#define GPID_SUPVAL 1
#define LAST_SERVER_SUPVAL 2
#define LAST_USER_SUPVAL 3
#define LASTUPDATE_SUPVAL 4
#define OPTIONS_SUPVAL 5

#define LASTUPDATE_ALTVAL 1
#define LAST_PORT_ALTVAL 2

#define CHALLENGE_SIZE 32
#define GPID_SIZE 32

#define MD5_LEN 16

#define NEW_PROJECT_INDEX 0

//User commands available via plugin activate, once connected to a server
#define USER_FORK       0
#define USER_CHECKPOINT 1
#define USER_PERMS      2
#define PROJECT_PERMS   3
#define USER_DISCONNECT 4
#define SHOW_NETNODE    5
#define CLEAN_NETNODE   6

extern bool publish;
extern bool userPublish;
extern bool subscribe;
extern bool supress;

extern bool authenticated;
extern bool fork_pending;

extern char **optLabels;

extern char username[64];

union OptVal {
   uint64_t ll;
   uint32_t ii[2];
};

struct Options {
   OptVal pub;
   OptVal sub;
};

extern Options userOpts;
extern Options tempOpts;
extern Options *optMasks;

extern uint64_t *snapUpdateIDs;
extern char description[MAXSTR];

extern int numProjectsGlobal;
extern int isSnapShotGlobal;
extern int numOptionsGlobal;

extern int *projects;

bool setUserOpts(Options &user);
bool getUserOpts(Options &user);

bool is_connected();
void cleanup(bool warn = false);
int send_all(const qstring &s);
int send_msg(const qstring &s);

bool init_network();
bool term_network();
#define QT_NAMESPACE QT

uint64_t getLastUpdate();
void setLastUpdate(uint64_t uid);
void writeUpdateValue(uint64_t uid);
void hookAll();
void unhookAll();
int numCommands();

void hmac_md5(unsigned char *msg, int msg_len,
              unsigned char *key, int key_len,
              unsigned char *digest);
void saveAuthData(char *user, char *pass);
int chooseProject(int index, char *desc);
bool changeProject(int index);
const char *getRunCommand(int i);
int getGpid(unsigned char *gpid, int sz);
void setGpid(unsigned char *gpid, int sz);
bool getFileMd5(unsigned char *md5, int len);
void do_project_rejoin();
void sendProjectLeave();
void do_project_leave();
void sendProjectChoice(int project);
void sendProjectSnapFork(int project, const char *desc);
void sendProjectGetList();
void sendNewProjectCreate(const char *description);
void sendReqPermsChoice();
void sendProjPermsChoice();
void freeProjectFields();
void selectProject(int index);
void sendAuthData(unsigned char *challenge, int challenge_len);
void do_get_req_perms(json_object *json);
void do_get_proj_perms(json_object *json);

bool do_choose_perms(json_object *json);

void do_send_user_message(const char *msg);

int send_json(json_object *obj);
int send_json(const char *type, json_object *obj);
int send_json(ea_t ea, const char *type, json_object *obj);

const char *hex_encode(const void *bin, uint32_t len);
uint8_t *hex_decode(const char *hex, uint32_t *len);
void format_llx(uint64_t val, qstring &s);

void append_json_hex_val(json_object *obj, const char *key, const uint8_t *value, uint32_t len = 0);
void append_json_string_val(json_object *obj, const char *key, const char *value);
void append_json_string_val(json_object *obj, const char *key, const qstring &value);
void append_json_bool_val(json_object *obj, const char *key, bool value);
void append_json_uint64_val(json_object *obj, const char *key, uint64_t value);
void append_json_uint32_val(json_object *obj, const char *key, uint32_t value);
void append_json_int32_val(json_object *obj, const char *key, int32_t value);
void append_json_ea_val(json_object *obj, const char *key, ea_t value);

uint8_t *hex_from_json(json_object *json, const char *key, uint32_t *len); //qfree this result
const char *string_from_json(json_object *json, const char *key);
bool bool_from_json(json_object *json, const char *key, bool *val);
bool uint64_from_json(json_object *json, const char *key, uint64_t *val);
bool ea_from_json(json_object *json, const char *key, ea_t *val);
bool uint32_from_json(json_object *json, const char *key, uint32_t *val);
bool int32_from_json(json_object *json, const char *key, int32_t *val);

//Ida's msg function doesn't handle the ll modifier
//this one returns pointer to static buffer
char *formatOptVal(OptVal *v);
char *formatLongLong(uint64_t);

//this one formats into buf, which should be at least 17 bytes in size
char *formatOptVal(OptVal *v, char *buf);
char *formatLongLong(uint64_t, char *buf);

void postCollabMessage(const char *msg, time_t t = 0);

//IDA HOOKS
#if IDA_SDK_VERSION < 700
int idaapi idb_hook(void * /*user_data*/, int notification_code, va_list va);
#else
ssize_t idaapi idb_hook(void * /*user_data*/, int notification_code, va_list va);
#endif
#if IDA_SDK_VERSION < 700
int idaapi idp_hook(void * /*user_data*/, int notification_code, va_list va);
//int idaapi ui_hook(void *user_data, int notification_code, va_list va);
#else
ssize_t idaapi idp_hook(void * /*user_data*/, int notification_code, va_list va);
#endif

//Collabreate messaging
void build_handler_map();
int handle_idp_msg(json_object *json, const char *msg_type);
bool msg_dispatcher(const char *json_in);

//help with the transition to IDA 7.0

#if IDA_SDK_VERSION >= 700

#define startEA start_ea
#define endEA end_ea

#define askbuttons_c ask_buttons
#define AskUsingForm_c ask_form
#define OpenForm_c open_form
#define FORM_TAB WOPN_TAB
#define FORM_QWIDGET 0

#define sup1st(n) supfirst(n)
#define supnxt(n, t) supnext(n, t)

#define get_flags_novalue(ea) get_flags(ea)
#define isEnum0(f) is_enum0(f)
#define isEnum1(f) is_enum1(f)
#define isStroff0(f) is_stroff0(f)
#define isStroff1(f) is_stroff1(f)
#define isOff0(f) is_off0(f)
#define isOff1(f) is_off1(f)
#define isOff(f, n) is_off(f, n)
#define isEnum(f, n) is_enum(f, n)
#define isStroff(f, n) is_stroff(f, n)

#define isStruct(f) is_struct(f)
#define isASCII(f) is_strlit(f)

#define get_member_name2 get_member_name

#else
// IDASDK_VERSION < 700

#define start_ea startEA
#define end_ea endEA

#define ask_buttons askbuttons_c
#define ask_form AskUsingForm_c
#define open_form OpenForm_c
#define WOPN_TAB FORM_TAB

#define supfirst(n) sup1st(n)
#define supnext(n, t) supnxt(n, t)

#define ev_add_cref add_cref
#define ev_add_dref add_dref
#define ev_del_cref del_cref
#define ev_del_dref del_dref
#define ev_auto_queue_empty auto_queue_empty
#define set_func_start func_setstart 
#define set_func_end func_setend

#endif

#endif
